src/compiler/program.ts

```diff
@@ -2004,15 +2023,25 @@ namespace ts {
 
                 Debug.assert(!!sourceFile.bindDiagnostics);
 
-                const isCheckJs = isCheckJsEnabledForFile(sourceFile, options);
+                const isJs = sourceFile.scriptKind === ScriptKind.JS || sourceFile.scriptKind === ScriptKind.JSX;
+                const isCheckJs = isJs && isCheckJsEnabledForFile(sourceFile, options);
+                const isPlainJs = isJs && !sourceFile.checkJsDirective && options.checkJs === undefined;

```

Comment:
What’s the difference between `isCheckJsEnabledForFile` the property checks you’re doing for `isPlainJs`?

Reply:
I *almost* inlined that function to make this easier to read, except that it's used everywhere. `isCheckJsEnabledForFile` is true whenever checkJs is explicitly true, whether locally or globally. To put it in the most parallel form:

`sourceFile.checkJsDirective?.enabled || options.checkJs === true` vs
`sourceFile.checkJsDirective === undefined && options.checkJs === undefined`

Reply:
Ah, so the third kind of JS is JS where `checkJs` is explicitly _disabled_?

Reply:
That's right. `// @ts-nocheck` is the local form of that.